perception/navigator_vision/nodes/pinger_finder.py

```diff
@@ -0,0 +1,170 @@
+point_distance = 10  # p1 is aling the direction of the pinger but we do not know the actual
+                     # range p1 will be adjusted to be 'point_distance' away from p0
+debug_arrow_id = 0
+listen = False
+
+# line_arr represents the equation of a 2d line. Structure: [x1, y1, x2, y2]
+# Calculates the point in the plane with the least cummulative distance to
+# every line in line_arr
+def LS_intersection(line_arr):
+    global pinger_position
+    def line_norm(line_end_pts):
+        assert len(line_end_pts) == 4
+        begin = line_end_pts[:2]
+        end = line_end_pts[2:]
+        diff = end - begin
+        return npl.norm(diff)
+
+    num_lines = line_arr.shape[0]
+    begin_pts = line_arr[:, :2]
+    end_pts = line_arr[:, 2:4]
+    diffs = end_pts - begin_pts
+    norms = np.apply_along_axis(line_norm, 1, line_arr)
+    norms = norms.reshape(norms.shape[0], 1)
+    perp_unit_vecs = np.apply_along_axis(lambda x: np.array([[0, -1], [1, 0]]).dot(x), 1, diffs / norms)
+    A_sum = np.zeros((2, 2))
+    Ap_sum = np.zeros((2, 1))
+
+    for x, y in zip(begin_pts, perp_unit_vecs):
+        begin = x.reshape(2, 1)
+        perp_vec = y.reshape(2, 1)
+        A = perp_vec.dot(perp_vec.T)
+        Ap = A.dot(begin)
+        A_sum += A
+        Ap_sum += Ap
+
+    res = npl.inv(A_sum).dot(Ap_sum)
+    pinger_position = Point(res[0], res[1], 0)
+
+
+def find_pinger(arg):
+    if line_array.shape[0] < 2:
+        return None
+    LS_intersection(line_array)
+    return {'pinger_position' : pinger_position, 'num_samples' : line_array.shape[0]}
+
+
+# receives pings and adds the representation of a line between the hydrophone array and the pinger
+# to our line_arr linear equation system
+def ping_cb(processed_ping):
+    global line_array, sample_amplitudes
+    print "Got processed ping message:\n{}".format(processed_ping)
+    if abs(processed_ping.freq - target_freq) < freq_tol and processed_ping.amplitude > min_amp:
+        print "\x1b[34mTrustworthy pinger heading\x1b[0m"
+        trans, rot = None, None
+        try:
+            tf_listener.waitForTransform("/enu", "/hydrophones", processed_ping.header.stamp, rospy.Duration(0.25))
+            trans, rot = tf_listener.lookupTransform("/enu", "/hydrophones", processed_ping.header.stamp)
+            p0 = np.array([trans[0], trans[1]])
+            R = tf.transformations.quaternion_matrix(rot)[:3, :3]
+            print "quat: {} rot mat: {}".format(rot, R)
+            print "delta hyd frame: {}".format(processed_ping.position)
+            delta = R.dot(navigator_tools.point_to_numpy(processed_ping.position))[:2]
+            print "delta enu frame: {}".format(delta)
+            p1 = p0 + point_distance * delta / npl.norm(delta)
+            line_coeffs = np.array([[p0[0], p0[1], p1[0], p1[1]]]) # p0 and p1 define a line
+            visualize_line(Point(p0[0], p0[1], 0), Point(p1[0], p1[1], 0))
+            if listen:
```

Comment:
Maybe ```if listen``` should be at the top of the callback. Shouldn't this node just idle if listen is false? This would save on the computational load when this is launched at startup. Also would provide more accurate markers.